group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/OptimizedUniformAssignmentBuilder.java

```diff
@@ -0,0 +1,399 @@
+     *
+     * @param memberId              The Id of the member whose assignment is being validated.
+     * @param assignedPartitions    The partitions currently assigned to the member.
+     *
+     * @return List of valid partitions after applying the filters.
+     */
+    private List<TopicIdPartition> getValidCurrentAssignment(String memberId, Map<Uuid, Set<Integer>> assignedPartitions) {
+        List<TopicIdPartition> validCurrentAssignmentList = new ArrayList<>();
+
+        assignedPartitions.forEach((topicId, currentAssignment) -> {
+            List<Integer> currentAssignmentList = new ArrayList<>(currentAssignment);
+            if (subscriptionList.contains(topicId)) {
+                for (Integer partition : currentAssignmentList) {
+                    TopicIdPartition topicIdPartition = new TopicIdPartition(topicId, partition);
+                    if (useRackAwareStrategy && rackInfo.racksMismatch(memberId, topicIdPartition)) {
+                        currentPartitionOwners.put(topicIdPartition, memberId);
+                    } else {
+                        validCurrentAssignmentList.add(topicIdPartition);
+                    }
+                }
+            }
+        });
+
+        return validCurrentAssignmentList;
+    }
+
+    /**
+     * This method iterates over the unassigned partitions and attempts to allocate them
+     * to members while considering their rack affiliations.
+     */
+    private void rackAwareRoundRobinAssignment() {
+        Queue<String> roundRobinMembers = new LinkedList<>(unfilledMembers.keySet());
+
+        // Sorts partitions in ascending order by number of potential consumers with matching racks.
+        // Partitions with no potential members aren't included in this list.
+        List<TopicIdPartition> sortedPartitions = rackInfo.sortPartitionsByRackConsumers(unassignedPartitions);
+
+        Iterator<TopicIdPartition> partitionIterator = sortedPartitions.iterator();
+        while (partitionIterator.hasNext()) {
+            TopicIdPartition partition = partitionIterator.next();
+            boolean assigned = false;
+            for (int i = 0; i < roundRobinMembers.size() && !assigned; i++) {
+                String memberId = roundRobinMembers.poll();
+                Integer remainingPartitionCount = unfilledMembers.get(memberId);
+
+                if (remainingPartitionCount != null && remainingPartitionCount > 0 && !rackInfo.racksMismatch(memberId, partition)) {
+                    assignPartitionToMember(memberId, partition);
+                    assigned = true;
+                    partitionIterator.remove();
+                    unassignedPartitions.remove(partition);
+                }
+
+                // Only re-add to the end of the queue if it's still in the unfilledMembers map
+                if (unfilledMembers.containsKey(memberId)) {
+                    roundRobinMembers.add(memberId);
+                }
```

Comment:
what i don't like about this approach is that the reader needs to know 2 things:
1. unfilledMembers (member -> remaining partition count) only exists when the count is > 0.
2. assignPartitionToMember removes a member from unfilledMembers if the count is 0 after decrementing

or else we don't know whether this for loop will end. also i have not seen a for loop where we add something to the original collection we are iterating through. i think `while (!queue.isEmpty())` should be more readable.

also, if there is a bug somewhere and for some reason the sum of all unfilledMembers` partition counts is greater than sortedPartitions size then we could have an infinite loop since there will be a member that still needs partitions to be assigned. we should add this check somewhere

Reply:
discussed offline that this isn't possible

Reply:
discussed offline that this isn't possible